packages/meditrak-app/app/api/TupaiaApi.js

```diff
@@ -230,59 +222,67 @@ export class TupaiaApi {
-      if (response.status === 410 && this.reduxStore) {
-        this.reduxStore.dispatch(logoutWithError(responseJson.error));
-      }
-      // If the server responded with any other error code but failed to supply an error, add one
-      if (
-        (response.status < 200 ||
-          response.status >= 300 ||
-          (responseJson.status && (responseJson.status < 200 || responseJson.status >= 300))) &&
-        !responseJson.error
-      ) {
-        responseJson.error = responseJson.message || `Network error ${response.status}`;
-
-        analytics.trackEvent('Request failed', {
-          requestMethod,
-          apiEndpoint,
-          error: responseJson.error,
-        });
-      } else {
-        analytics.trackEvent('Request succeeded', {
-          requestMethod,
-          apiEndpoint,
-        });
-      }
-      return responseJson;
-    } catch (error) {
-      throw error;
+    const response = await fetchWithTimeout(queryUrl, fetchConfig);
+    // If server responded with 401, i.e. not authenticated, refresh token and try once more
+    if (
+      shouldReauthenticateIfUnauthorized &&
+      response.status === 401 &&
+      this.refreshToken !== null
+    ) {
+      await this.refreshAccessToken();
+      return this.request(requestMethod, apiEndpoint, queryParameters, body, null, false);
     }
+
+    const responseJson = await response.json();
+    // If server responded with 410, i.e. old API version, log the user out and get them to update
+    if (response.status === 410 && this.reduxStore) {
+      this.reduxStore.dispatch(logoutWithError(responseJson.error));
+    }
+    // If the server responded with any other error code but failed to supply an error, add one
+    if (
+      (response.status < 200 ||
+        response.status >= 300 ||
+        (responseJson.status && (responseJson.status < 200 || responseJson.status >= 300))) &&
+      !responseJson.error
+    ) {
+      responseJson.error = responseJson.message || `Network error ${response.status}`;
+
+      analytics.trackEvent('Request failed', {
+        requestMethod,
+        apiEndpoint,
+        error: responseJson.error,
+      });
+    } else {
+      analytics.trackEvent('Request succeeded', {
+        requestMethod,
+        apiEndpoint,
+      });
+    }
+    return responseJson;
   }
 }
 
 // Create a promise that rejects after the request has taken too long
-const createTimeoutPromise = () =>
-  new Promise((resolve, reject) => {
+const createTimeoutPromise = () => {
+  let cleanup;
+  const promise = new Promise((resolve, reject) => {
     const id = setTimeout(() => {
       clearTimeout(id);
-      reject({ message: 'Network request timed out' });
+      reject(new Error('Network request timed out'));
     }, TIMEOUT_INTERVAL);
+    cleanup = () => {
+      clearTimeout(id);
+      resolve();
+    };
   });
+  return { promise, cleanup };
+};
+export const fetchWithTimeout = async (url, config) => {
```

Comment:
This function seems identical to the one in `@tupaia/utils`... I guess we have reasons for not including that package in `meditrak-app`?

Reply:
Yes unfortunately it breaks things, because it imports all this node stuff that's not supported in the react-native environment